### DISCUSSION TITLE
Carpet rules

### DISCUSSION THREAD

**Discussion by object-Object** [HexxyCraft/modpack#9] [General]
What [Carpet](https://modrinth.com/mod/carpet) rules/configs should we use?

**Comment by LlamaSageR** [comment]
The primary utility for Carpet is the op commands for manipulating tick speed and profiling to help find and resolve bugs on the server. Renewable Blackstone, Coral, and Deepslate is a nice addition though. (Renewable Sponges is nice on paper, but being able to convert Guardians into Elder Guardians can only end poorly. Still up for consideration though)

Moveable Tile Entities is also an option, not sure how well that functions though.

**Comment by beholderface** [comment]
- antiCheatDisabled - would allow for more freedom with stuff like custom hex flight solutions
- chainStone - cool directional pseudo-slimeblock
- commandProfile - can help to diagnose lag in your base
- ctrlQCraftingFix - mass crafting things go brr
- lightningKillsDropsFix - become god of thunder while still collecting mob drops
- movableBlockEntities - fun redstone shenanigans 
- optimizedTNT - less lag for hopefully the same effect is almost always good
- renewableBlackstone, renewableCoral, renewableDeepslate, renewableSponges - renewability go brr

**Reply by autumn-birds** [reply]
Does movableBlockEntities let you relocate spawners with a redstone flying machine?  Which, to be fair, I don't think that that would be the _worst_ thing in the world, but I am curious.

**Reply by beholderface** [reply]
Maybe? But I'm pretty sure you can do it super easily with Create (just assemble a minecart contraption with the spawner on it), so it's not like we couldn't move spawners without Carpet. And IIRC Botania has something specifically for moving spawners.

**Reply by autumn-birds** [reply]
Yeah i don't think it's a bad thing. It was mainly "hmm, i wonder if ..."

**Comment by object-Object** [comment]
Thoughts on these?
* Wither rose chunkloaders: balanced vanilla-y chunkloading instead of just a menu in your inventory
* [/playerme](https://github.com/gnembon/scarpet/blob/master/programs/survival/playerme.sc) + [keepalive](https://github.com/gnembon/scarpet/blob/master/programs/utilities/keepalive.sc): fake players that only work if you're targeting yourself, for automated afk when you're offline
* [/cam](https://github.com/gnembon/scarpet/blob/master/programs/survival/cam.sc): safe/fair spectator mode

**Comment by beholderface** [comment]
The wither rose chunkloaders seem like they could pretty easily result in a lot of lag, with how it's not too difficult to make wither rose farms IIRC. KyanBirb posted [a mod](https://modrinth.com/mod/chunk-loaders) in the suggestions discord thread that has chunkloader blocks which can be configured so that any given player can only have so many active, maybe use that? (I'm adding it to the spreadsheet now)
Edit: looks like it doesn't have a 1.19.2 Fabric version, dang

I admittedly don't know much about the capabilities of Carpet's fake player stuff, but couldn't they present worse server performance issues than having unlimited chunk loaders?

Even with the restrictions of not being able to use it for travel or escaping fights and such, I think access to spectator mode would be too easily abused for purposes like finding lots of otherwise-rare structures, or close-together geodes that all have slipways, that sort of thing.

**Comment by LlamaSageR** [comment]
I concur. While having an in-world chunkloader is good, having it be centered around Wither Roses would make it trivial for players to chunkload vast areas, resulting in incredible lag.

Same goes for the fake player command, we don't really need a player to automate most things. Carpet's utilities are largely focused around Vanilla, so the fake player to AFK is likely an aspect of that. A simple chunkloader would resolve this issue.

Hard pass on the spectator mode. Trivializes exploration and is way too easy to abuse.

CommandProfile is likely covered by Spark, so not sure how necessary that is. It can be enabled if we do run into issues though, since that's more of a server-side config.

**Comment by object-Object** [comment]
I thought the fake players worked quite well on Arcane Compendium. They didn't seem to cause lag afaik. They allow automatic xp farms, fortune, etc, without having to actually leave your computer connected. That's probably actually less laggy because it's entirely server-side at that point. Also, I'm pretty sure most chunkloaders, unlike fake players, don't do random ticks.

**Comment by LlamaSageR** [comment]
The main cause of lag on servers is people keeping massive bases perpetually loaded, which is why Chunkloading is limited and AFK players are disconnected after ~30 minutes. Having a tool that allows players to entirely bypass those limitations will cause issues with lag later down the line. Not to mention that XP is less necessary with the repair changes (Mending is no longer mandatory and costs are reduced across the board) and Create's Deployers or CC Turtles can use Fortune. 
This is not even addressing the possible ramifications of having server-endorsed bots that are perpetually relogged if the server restarts.

**Reply by object-Object** [reply]
> AFK players are disconnected after ~30 minutes

What??? Absolutely not. When was this discussed? That's a huge, unnecessary, and annoying premature optimization. If this is in the pack, please remove it. *If* players being AFK actually ends up causing problems, we can *discuss* re-adding this.

> having server-endorsed bots that are perpetually relogged if the server restarts

Okay yeah that's a good point.

**Reply by object-Object** [reply]
However:
* `/playerme` is a tweak that we used for most of Arcane Compendium with, as far as I know, zero problems. It's battle-tested. And in the worst case scenario, if it *does* cause problems, it's easy to remove.
* If the thing causing lag is the massive bases, then limit that instead of the AFK tools. For example, do we have mods that mitigate the need for massive animal farms? If not, we probably should.
* ~~As I said before, chunkloaders don't cause random ticks (eg. for crop and amethyst growth) as far as I know. Is there another solution for this?~~ Depends on the mod apparently. FTB Chunks has a config to enable random ticks in force loaded chunks, for example.
* People are going to AFK at farms. It's part of the game, and we shouldn't limit it. This is just a tool to make it more ergonomic for people to do something that they'll be doing anyway.

**Reply by LlamaSageR** [reply]
Oh, I was referring to other servers in general, that part must've been cut out while editing. There's absolutely no feature to kick AFK-Players right now and no need to introduce that. Sorry for the confusion lol.

**Reply by LlamaSageR** [reply]
I feel /playerme disincentivizes more natural automation, with Create and Hex Casting and whatnot. It feels like a very blunt tool for solving a problem that isn't really present. Let me address your points

- I don't doubt that /playerme (how do you do that formatting btw?) is functional and nonbuggy. I'd prefer not to give players a tool and then have to remove it.

- The issue is that any base that's large enough will cause more lag, regardless of what tools we add the alleviate it. The best way to alleviate base lag is to simply not have it be loaded 24/7 rather than optimizing the performance of the base.

- Yeah, depends on the mod. Don't have much to add here lol.

- That's the main thing though, people don't *need* to AFK their farms. With chunkloaders and whatnot, the player doesn't have to be online for them to run, and there aren't really any farms that need players exclusively (at least that I can think of, or are that vital). People will go AFK, I will grant that. But there's no need to incentivize it. No need to *limit* AFK, yes, but there's also no need to *promote it*.

And none of this is to mention the potential shenanigans of it being actual bots as opposed to a method of just going AFK.

**Reply by object-Object** [reply]
I think I agree with this now and am willing to concede. Thanks for your points. If the server gets going and a lot of people end up missing `/playerme`, we can always discuss this again.

Formatting: \`/playerme\` (same as on Discord)

**Reply by LlamaSageR** [reply]
Gotcha gotcha. And thank you for the formatting help lol

**Reply by leo60228** [reply]
`/playerme` would allow automating the Gaia Guardian, for better or for worse.

**Reply by LlamaSageR** [reply]
That's... probably for the worse in all honesty.

**Comment by leo60228** [comment]
Considering the Hexal falling block spell, it might be worth adding [Carpet-Fixes](https://modrinth.com/mod/carpet-fixes) and enabling (at least) `fallingBlockDuplicationFix`.

**Reply by LlamaSageR** [reply]
Yeah, Carpet-Fixes is definitely going on the list. Lots of good bugfixing and such.
